devito/function.py

```diff
@@ -362,30 +365,33 @@ def __init__(self, *args, **kwargs):
         if not self._cached():
             super(TimeFunction, self).__init__(*args, **kwargs)
             self.time_dim = kwargs.get('time_dim', None)
+            if self.time_dim is not None and not isinstance(self.time_dim, TimeDimension):
+                raise ValueError("time_dim must be a TimeDimension, not %s" %
+                                 type(self.time_dim))
             self.time_order = kwargs.get('time_order', 1)
-            self.save = kwargs.get('save', False)
-
-            if not self.save:
-                if self.time_dim is not None:
-                    warning('Explicit time dimension size (time_dim) found for '
-                            'TimeFunction symbol %s, despite \nusing a stepping time '
-                            'dimension (save=False). This value will be ignored!'
-                            % self.name)
-                self.time_dim = self.time_order + 1
-                self.indices[0].modulo = self.time_dim
+            self.save = kwargs.get('save', None)
+
+            if self.save is not None:
+                if not isinstance(self.save, int):
+                    raise ValueError("save must be an int indicating the number of " +
+                                     "timesteps to be saved (is %s)" % type(self.save))
+                available_mem = virtual_memory().available
+
+                if np.dtype(self.dtype).itemsize * self.save > available_mem:
+                    warning("Trying to allocate more memory for symbol %s " % self.name +
+                            "than available on physical device, this will start swapping")
```

Comment:
crazy thought that we probably do not wanna do, but maybe it's worth talking about it... 
How about we run ``clean_cache()`` here? so that "this might start swapping", rather than "this will start swapping". Might help with outer loops allocating new wavefields at each iteration, whose references are kept by the sympy cache.

If we did that, we should perform this check regardless of ``self.save``

Reply:
`clean_cache()` is potentially a heavy operation invoking the GC - not something I'll be comfortable doing without making it clear to the user (who might be multiple application-abstraction layers above devito). Plus, it is required to cover for a "bug" in devito.